File: frontend/src/kakao-login.ts
import type { Session, SupabaseClient } from "@supabase/supabase-js";
import { apiUrl } from "./api";

const PENDING_KEY = "jangbu:kakao-login";
const MAX_PENDING_AGE_MS = 10 * 60 * 1000;

interface PendingLogin {
  state: string;
  nonce: string;
  codeVerifier: string;
  createdAt: number;
}

function randomHex(): string {
  const bytes = crypto.getRandomValues(new Uint8Array(32));
  return Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0")).join("");
}

async function sha256(value: string): Promise<string> {
  const hash = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(value));
  return Array.from(new Uint8Array(hash), (byte) => byte.toString(16).padStart(2, "0")).join("");
}

async function pkceChallenge(verifier: string): Promise<string> {
  const bytes = new Uint8Array(await crypto.subtle.digest("SHA-256", new TextEncoder().encode(verifier)));
  return btoa(String.fromCharCode(...bytes)).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

async function post<T>(path: string, body: Record<string, string>): Promise<T> {
  const response = await fetch(apiUrl(path), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
    cache: "no-store",
  });
  if (!response.ok) throw new Error("kakao_request_failed");
  const payload = await response.json() as { data?: T };
  if (!payload.data) throw new Error("kakao_response_invalid");
  return payload.data;
}

export function isKakaoCallback(): boolean {
  return window.location.pathname === "/auth/kakao";
}

export async function startKakaoLogin(): Promise<void> {
  const pending: PendingLogin = {
    state: randomHex(),
    nonce: randomHex(),
    codeVerifier: randomHex(),
    createdAt: Date.now(),
  };
  const { authorization_url } = await post<{ authorization_url: string }>(
    "/api/v1/auth/kakao/start",
    {
      state: pending.state,
      nonce_hash: await sha256(pending.nonce),
      code_challenge: await pkceChallenge(pending.codeVerifier),
    },
  );
  const destination = new URL(authorization_url);
  if (destination.origin !== "https://kauth.kakao.com" || destination.pathname !== "/oauth/authorize") {
    throw new Error("kakao_authorization_url_invalid");
  }
  sessionStorage.setItem(PENDING_KEY, JSON.stringify(pending));
  window.location.assign(destination.href);
}

export async function finishKakaoLogin(supabase: SupabaseClient): Promise<Session> {
  const query = new URLSearchParams(window.location.search);
  const saved = sessionStorage.getItem(PENDING_KEY);
  sessionStorage.removeItem(PENDING_KEY);
  window.history.replaceState(null, "", "/");
  let pending: PendingLogin | null = null;
  try {
    pending = saved ? JSON.parse(saved) as PendingLogin : null;
  } catch {
    // Treat a damaged login attempt as invalid.
  }
  if (!pending || !pending.state || !pending.nonce || !pending.codeVerifier
    || !Number.isFinite(pending.createdAt)
    || Date.now() - pending.createdAt > MAX_PENDING_AGE_MS
    || Date.now() < pending.createdAt
    || query.get("state") !== pending.state) {
    throw new Error("kakao_state_invalid");
  }
  if (query.has("error")) throw new Error("kakao_authorization_denied");
  const code = query.get("code");
  if (!code) throw new Error("kakao_code_missing");
  const { id_token } = await post<{ id_token: string }>("/api/v1/auth/kakao/exchange", {
    code,
    code_verifier: pending.codeVerifier,
  });
  const { data, error } = await supabase.auth.signInWithIdToken({
    provider: "kakao",
    token: id_token,
    nonce: pending.nonce,
  });
  if (error || !data.session) throw error ?? new Error("kakao_session_missing");
  return data.session;
}

File: ledger/kakao_login.py
from __future__ import annotations

import re
from urllib.parse import urlencode

import requests


class KakaoLoginError(Exception):
    pass


_STATE = re.compile(r"^[A-Za-z0-9_-]{32,128}$")
_NONCE_HASH = re.compile(r"^[0-9a-f]{64}$")
_CODE = re.compile(r"^[\x21-\x7e]{1,1024}$")
_PKCE_CHALLENGE = re.compile(r"^[A-Za-z0-9_-]{43}$")
_PKCE_VERIFIER = re.compile(r"^[A-Za-z0-9_-]{43,128}$")


def authorization_url(
    client_id: str, origin: str, state: str, nonce_hash: str, code_challenge: str
) -> str:
    if (not client_id or not _STATE.fullmatch(state) or not _NONCE_HASH.fullmatch(nonce_hash)
            or not _PKCE_CHALLENGE.fullmatch(code_challenge)):
        raise KakaoLoginError("invalid Kakao login request")
    query = urlencode({
        "response_type": "code",
        "client_id": client_id,
        "redirect_uri": f"{origin}/auth/kakao",
        "state": state,
        "nonce": nonce_hash,
        "code_challenge": code_challenge,
        "code_challenge_method": "S256",
    })
    return f"https://kauth.kakao.com/oauth/authorize?{query}"


def exchange_code(
    client_id: str, client_secret: str, origin: str, code: str, code_verifier: str
) -> str:
    if not client_id or not client_secret:
        raise KakaoLoginError("Kakao login is not configured")
    if (not isinstance(code, str) or not _CODE.fullmatch(code)
            or not isinstance(code_verifier, str) or not _PKCE_VERIFIER.fullmatch(code_verifier)):
        raise KakaoLoginError("invalid Kakao authorization code")
    try:
        response = requests.post(
            "https://kauth.kakao.com/oauth/token",
            data={
                "grant_type": "authorization_code",
                "client_id": client_id,
                "client_secret": client_secret,
                "redirect_uri": f"{origin}/auth/kakao",
                "code": code,
                "code_verifier": code_verifier,
            },
            headers={"Content-Type": "application/x-www-form-urlencoded;charset=utf-8"},
            timeout=10,
            allow_redirects=False,
        )
        response.raise_for_status()
        payload = response.json()
    except (requests.RequestException, ValueError) as error:
        raise KakaoLoginError("Kakao token exchange failed") from error
    token = payload.get("id_token") if isinstance(payload, dict) else None
    if not isinstance(token, str) or not token:
        raise KakaoLoginError("Kakao did not issue an ID token")
    return token
